modules/apigee_edge_apidocs/src/ApiDocSpecFetcher.php

```diff
@@ -0,0 +1,264 @@
+   *
+   * @var \GuzzleHttp\ClientInterface
+   */
+  protected $httpClient;
+
+  /**
+   * The entity type manager.
+   *
+   * @var \Drupal\Core\Entity\EntityTypeManagerInterface
+   */
+  protected $entityTypeManager;
+
+  /**
+   * Drupal\Core\Messenger\MessengerInterface definition.
+   *
+   * @var \Drupal\Core\Messenger\MessengerInterface
+   */
+  protected $messenger;
+
+  /**
+   * The logger.
+   *
+   * @var \Psr\Log\LoggerInterface
+   */
+  private $logger;
+
+  /**
+   * Constructs a new ApiDocSpecFetcher object.
+   *
+   * @param \Drupal\Core\File\FileSystemInterface $file_system
+   *   The file_system service.
+   * @param \GuzzleHttp\ClientInterface $http_client
+   *   The http_client service.
+   * @param \Drupal\Core\Entity\EntityTypeManagerInterface $entityTypeManager
+   *   The entity type manager.
+   * @param \Drupal\Core\Messenger\MessengerInterface $messenger
+   *   The messenger service.
+   * @param \Psr\Log\LoggerInterface $logger
+   *   The logger service.
+   */
+  public function __construct(FileSystemInterface $file_system, ClientInterface $http_client, EntityTypeManagerInterface $entityTypeManager, MessengerInterface $messenger, LoggerInterface $logger) {
+    $this->fileSystem = $file_system;
+    $this->httpClient = $http_client;
+    $this->entityTypeManager = $entityTypeManager;
+    $this->messenger = $messenger;
+    $this->logger = $logger;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function fetchSpec(ApiDocInterface $apidoc, bool $show_messages = TRUE): bool {
+    $needs_save = FALSE;
+    $spec_value = $apidoc->get('spec')->isEmpty() ? [] : $apidoc->get('spec')->getValue()[0];
+
+    // If "spec_file_source" uses URL, grab file from "file_link" and save it
+    // into the "spec" file field. The file_link field should already have
+    // validated that a valid file exists at that URL.
+    if ($apidoc->get('spec_file_source')->value == ApiDocInterface::SPEC_AS_URL) {
+
+      // If the file_link field is empty, return without changes.
+      if ($apidoc->get('file_link')->isEmpty()) {
+        return FALSE;
+      }
+
+      $file_uri = $apidoc->get('file_link')->getValue()[0]['uri'];
+      $file_uri = Url::fromUri($file_uri, ['absolute' => TRUE])->toString();
+      $request = new Request('GET', $file_uri);
+      $options = [
+        'exceptions' => TRUE,
+        'allow_redirects' => [
+          'strict' => TRUE,
+        ],
+      ];
+
+      // Generate conditional GET header.
+      if (!$apidoc->get('fetched_timestamp')->isEmpty()) {
+        $request = $request->withAddedHeader('If-Modified-Since', gmdate(DateTimePlus::RFC7231, $apidoc->get('fetched_timestamp')->value));
+      }
+
+      try {
+        $response = $this->httpClient->send($request, $options);
+
+        // In case of a 304 Not Modified there are no changes, but update
+        // last fetched timestamp.
+        if ($response->getStatusCode() == 304) {
+          $apidoc->set('fetched_timestamp', time());
```

Comment:
Use `\Drupal\Component\Datetime\Time::getCurrentTime()` (from the injeced service: `datetime.time`) instead as per: https://www.drupal.org/node/2785211

Reply:
`time() !== REQUEST_TIME`. Why would it be necessary to go through `\Drupal\Component\Datetime\Time::getCurrentTime()` since it just calls the native PHP function.

https://api.drupal.org/api/drupal/core!lib!Drupal!Component!Datetime!Time.php/class/Time/8.7.x

Reply:
Yes, I checked this one (not like the RFC7231 ;P) , injecting time as a service (as a mockable PHP object) allows to mock time while you are testing the API doc fetcher service :)